### ISSUE TITLE
SDPA reads zero-padded mask columns when Sk % 32 != 0

### ISSUE THREAD

**Issue by lhl** [tenstorrent/tt-metal#54833]
## Summary

On Wormhole with tt-metal `66a55a6fbee`,
`ttnn.transformer.scaled_dot_product_attention` returned incorrect BF16 output
under these conditions:

- we supplied an explicit BF16 `attn_mask`;
- we created the mask with `ttnn.from_torch`, which pads TILE tensors with zero
  by default; and
- the key sequence length was not a multiple of the 32-element tile width. We
  tested S=16 and S=40.

Changing only the mask's physical padding to negative infinity fixed the
result:

```python
pad_value=float("-inf")
```

At S=16 and S=40, the fixed output was bit-for-bit identical to the accurate
`is_causal=True` device control. At the tile-aligned length S=32, both padding
values produced the correct output. At S=16, repeated calls confirmed that
both the failure and the fix were deterministic.

The results also separate physical padding from the mask's logical values. At
S=16, S=32, and S=40, logical upper-triangle values of `-1e9` and `-inf`
produced bit-for-bit identical output. Both values failed when the physical
padding remained zero. This shows that SDPA reads the zero-valued columns in
the final partial mask tile.

## Why this matters

PR #36170 documents support for provided masks at arbitrary non-chunk-aligned
Q/K lengths. A caller following that contract can receive a well-formed tensor
with severely incorrect values when `ttnn.from_torch` uses its default tile
padding. Applications need explicit masks for behavior that
`is_causal=True` cannot express, including additive bias, padding, and token
selection. We found this bug while implementing sparse-selection attention.

## Environment

- tt-metal / TTNN: `66a55a6fbee8bccba60a8225bd10aece6f62e556`
  (`v0.76.0-dev20260728-18-g66a55a6fbee`)
- Firmware bundle: `19.13.1`
- Kernel driver: `2.10.0`
- Hardware: Wormhole Galaxy, one host-facing ASIC at `0000:c6:00.0`, PCIe
  Gen4 x8
- Isolation: exclusive lease, `TT_VISIBLE_DEVICES=29`, process-local device 0
- Python: CPython `3.10.19`
- Q/K/V and mask: BF16 TILE layout
- Shape: Q `[1,24,S,256]`; K/V `[1,2,S,256]`; mask `[1,1,S,S]`
- Scale: `1/sqrt(256) = 0.0625`
- Seed: `20260829`
- Reference: PyTorch MATH SDPA with BF16-rounded Q/K/V and BF16-rounded output

Before running the matrix, we reset only `0000:c6:00.0` with
`tt-smi -r`. The reset completed with exit code 0. We then ran an isolated BF16
matmul control, which passed at PCC `0.999988995`.

The SDPA test opened and closed the device cleanly. Its log contained no
unexpected-mailbox, failed-early-exit, stale-dispatch, or
failed-firmware-initialization markers. Before and after the test, `c6`
reported healthy DRAM, live Ethernet, faults `0x0`, throttler `0x0`, and a
temperature range of 26.4–28.0 C.

## Clean test results

The table reports the Pearson correlation coefficient (PCC), root mean square
error (RMSE), and maximum absolute error against the PyTorch reference.

| S | Case | PCC vs PyTorch | RMSE | max abs | Output comparison |
| ---: | --- | ---: | ---: | ---: | --- |
| 16 | no mask | 0.999777249 | 0.0088650 | 0.06640625 | full-attention control |
| 16 | `is_causal=True` | 0.999836732 | 0.0130208 | 0.109375 | causal device control |
| 16 | provided mask, default zero padding | **0.846577666** | **0.3857387** | **3.298828125** | corrupted |
| 16 | provided mask, `pad_value=-inf` | 0.999836732 | 0.0130208 | 0.109375 | **bit-for-bit equal to causal control** |
| 16 | pad Q/K/V/mask to S=32 | 0.999836732 | 0.0130208 | 0.109375 | **bit-for-bit equal for retained rows** |
| 32 | `is_causal=True` | 0.999821368 | 0.0105774 | 0.125 | causal device control |
| 32 | provided mask, default zero padding | 0.999821368 | 0.0105774 | 0.125 | **bit-for-bit equal to causal control** |
| 32 | provided mask, `pad_value=-inf` | 0.999821368 | 0.0105774 | 0.125 | **bit-for-bit equal to causal control** |
| 40 | no mask | 0.999707378 | 0.0068025 | 0.0703125 | full-attention control |
| 40 | `is_causal=True` | 0.999816218 | 0.0104603 | 0.125 | causal device control |
| 40 | provided mask, default zero padding | **0.831548585** | **0.2966589** | **3.808837891** | corrupted |
| 40 | provided mask, `pad_value=-inf` | 0.999816218 | 0.0104603 | 0.125 | **bit-for-bit equal to causal control** |

We compared the complete output hashes to verify every bit-for-bit equality in
the table.

Earlier diagnostic runs showed the same alignment pattern at S=48 (failed) and
S=64 (passed). The clean matrix above is the evidence used for this report. It
covers one seed, one chip, BF16, square self-attention, native grouped-query
attention, and S=16/32/40. We have not tested every non-tile-aligned shape or
every supported mask dtype.

## Minimal reproduction

Isolate the target physical device before importing TTNN. Our physical UMD ID
was 29; use the appropriate ID on another system. Inside the process, the
isolated device is device 0.

```bash
TT_VISIBLE_DEVICES=29 \
TT_METAL_CACHE=/tmp/tt-rt-004-cache \
python repro.py
```

```python
import math
import torch
from torch.nn.attention import SDPBackend, sdpa_kernel
import ttnn

NQH, NKV, DH, S = 24, 2, 256, 16


def pcc(a, b):
    a = a.flatten().double()
    b = b.flatten().double()
    a -= a.mean()
    b -= b.mean()
    return float((a * b).sum() / (a.norm() * b.norm()).clamp_min(1e-12))


def upload(t, device, pad_value=None):
    kwargs = dict(dtype=ttnn.bfloat16, layout=ttnn.TILE_LAYOUT, device=device)
    if pad_value is not None:
        kwargs["pad_value"] = pad_value
    return ttnn.from_torch(t, **kwargs)


torch.manual_seed(20260829)
q = torch.randn(1, NQH, S, DH).to(torch.bfloat16).float()
k = torch.randn(1, NKV, S, DH).to(torch.bfloat16).float()
v = torch.randn(1, NKV, S, DH).to(torch.bfloat16).float()
mask = torch.zeros(S, S)
mask.masked_fill_(torch.ones(S, S, dtype=torch.bool).triu(1), -1e9)
mask = mask.unsqueeze(0).unsqueeze(0)
scale = 1.0 / math.sqrt(DH)

with sdpa_kernel(SDPBackend.MATH):
    reference = torch.nn.functional.scaled_dot_product_attention(
        q,
        k.repeat_interleave(NQH // NKV, dim=1),
        v.repeat_interleave(NQH // NKV, dim=1),
        is_causal=True,
        scale=scale,
    ).to(torch.bfloat16).float()

device = ttnn.open_device(device_id=0)
try:
    qt, kt, vt = upload(q, device), upload(k, device), upload(v, device)

    def run(mask_tt=None, is_causal=False):
        out = ttnn.transformer.scaled_dot_product_attention(
            qt, kt, vt, attn_mask=mask_tt, is_causal=is_causal, scale=scale
        )
        return ttnn.to_torch(out).float()

    causal = run(is_causal=True)
    broken = run(upload(mask, device))
    repaired = run(upload(mask, device, pad_value=float("-inf")))

    print("causal PCC:", pcc(causal, reference))
    print("default-padding PCC:", pcc(broken, reference))
    print("-inf-padding PCC:", pcc(repaired, reference))
    print("repaired equals causal:", torch.equal(repaired, causal))
finally:
    ttnn.close_device(device)
```

Expected output on the affected build is approximately:

```text
causal PCC: 0.99984
default-padding PCC: 0.85
-inf-padding PCC: 0.99984
repaired equals causal: True
```

## Relevant source and upstream changes

This appears to be a missing partial-tile case in the arbitrary-length
provided-mask support requested in
[issue #36097](https://github.com/tenstorrent/tt-metal/issues/36097) and added
by [PR #36170](https://github.com/tenstorrent/tt-metal/pull/36170). PR #36170
says that the reader fills padded Q and K mask positions with negative
infinity. Its S=160 test is not chunk-aligned in that test configuration, but
160 is still a multiple of the 32-element tile width. The test therefore does
not cover a partially filled final tile.

[PR #46600](https://github.com/tenstorrent/tt-metal/pull/46600) later moved
provided dense masks to streaming SDPA. Its reported K-padding tests ran on
Blackhole and do not state whether they covered a partial final tile on
Wormhole.

At the commit we tested, the code handles the mask as follows:

- [`py_to_tt_tensor.cpp`](https://github.com/tenstorrent/tt-metal/blob/66a55a6fbee8bccba60a8225bd10aece6f62e556/ttnn/core/tensor/py_to_tt_tensor.cpp#L442-L454)
  uses `pad_value.value_or(0.0f)`, so a tiled mask has zero-valued physical
  padding unless the caller overrides it.
- [`reader_interleaved.cpp`](https://github.com/tenstorrent/tt-metal/blob/66a55a6fbee8bccba60a8225bd10aece6f62e556/ttnn/cpp/ttnn/operations/transformer/sdpa/device/kernels/dataflow/reader_interleaved.cpp#L521-L556)
  tests `global_k_tile < valid_Skt`. The final partial K tile is therefore
  considered valid and copied in full. Only wholly out-of-range tiles are
  replaced by `fill_neginf_tile`.
- [`sdpa_program_factory.cpp`](https://github.com/tenstorrent/tt-metal/blob/66a55a6fbee8bccba60a8225bd10aece6f62e556/ttnn/cpp/ttnn/operations/transformer/sdpa/device/sdpa_program_factory.cpp#L419-L428)
  enables its partial-column mask only for
  `generated_padding_mask && !use_provided_mask`. The adjacent comment says the
  reader fills padded positions for a dense provided mask, but the reader's
  decision is at whole-tile granularity.

This is not the mask-scaling bug fixed by
[PR #42891](https://github.com/tenstorrent/tt-metal/pull/42891): that change is
an ancestor of the tested build, aligned S works at the same scale, and the
failure is repaired solely by changing physical padding. It is also not one of
the five SDPA defects grouped in
[issue #51223](https://github.com/tenstorrent/tt-metal/issues/51223).

## Expected behavior

Values outside a mask's logical S×S shape should not change the SDPA output. If
provided masks with `Sk % 32 != 0` are unsupported, SDPA should reject them
instead of returning a well-formed tensor with incorrect values.

## Verified workarounds

The simplest workaround on the affected build is to initialize the mask's
physical padding to negative infinity:

```python
mask_tt = ttnn.from_torch(
    mask,
    dtype=ttnn.bfloat16,
    layout=ttnn.TILE_LAYOUT,
    device=device,
    pad_value=float("-inf"),
)
```

Padding Q/K/V and the mask to the next multiple of 32 also works, but it
changes the tensor shapes and performs unnecessary work. A causal-only path can
avoid a provided mask and use `is_causal=True`.
